lib/mp4/transmuxer.js

```diff
@@ -179,90 +259,76 @@ VideoSegmentStream = function(track) {
       nalUnits.shift();
     }
 
-    // return early if no video data has been observed
-    if (nalUnitsLength === 0) {
+    // Return early if no video data has been observed
+    if (nalUnits.length === 0) {
+      this.resetStreamStage_();
       this.trigger('done');
       return;
     }
 
-    // concatenate the video data and construct the mdat
-    // first, we have to build the index from byte locations to
-    // samples (that is, frames) in the video data
-    data = new Uint8Array(nalUnitsLength + (4 * nalUnits.length));
-    view = new DataView(data.buffer);
-    track.samples = [];
-
-    // see ISO/IEC 14496-12:2012, section 8.6.4.3
-    sample = {
-      size: 0,
-      flags: {
-        isLeading: 0,
-        dependsOn: 1,
-        isDependedOn: 0,
-        hasRedundancy: 0,
-        degradationPriority: 0
+    // Organize the raw nal-units into arrays that represent
+    // higher-level constructs such as frames and gops
+    // (group-of-pictures)
+    frames = this.groupNalsIntoFrames_(nalUnits);
+    gops = this.groupFramesIntoGops_(frames);
+
+    // If the first frame of this fragment is not an i-frame we have
+    // a problem since MSE (on Chrome) requires a leading i-frame.
+    //
+    // We have two approaches to repairing this situation:
+    // 1) GOP-FUSION:
+    //    This is where we keep track of the GOPS (group-of-pictures)
+    //    from previous fragments and attempt to find one that we can
+    //    prepend to the current fragment in order to create a valid
+    //    fragment.
+    // 2) I-FRAME-PULLING:
+    //    Here we search for the first i-frame in the fragment and
+    //    throw away all the frames between the start of the fragment
+    //    and that i-frame. We then extend the duration and pull the
+    //    PTS of the i-frame forward so that it covers the time range
+    //    of the frames that were disposed of.
+    //
+    // #1 is far prefereable over #2 which can cause "studdering" but
+    // requires more things to be just right.
+    if (!gops[0][0].isIDR) {
```

Comment:
I believe this should be `isIdr` by our capitalization rules. I also think we need to be much more consistent with terminology. The FLV transmuxer has a boolean `keyFrame` on frame objects. You also refer to this as an "i-frame" in your comments above and don't mention IDRs anywhere. Let's choose one and stick with it.